### PR TITLE
Add documentation for `any_resource`

### PR DESCRIPTION
**Pull request by miscco**
The docs weren building properly, so add some

### CODE REVIEW

File: cudax/include/cuda/experimental/__memory_resource/any_resource.h [hunk 1]
```diff
@@ -95,10 +105,12 @@ class basic_any_resource
     }
   }
 
-#ifndef DOXYGEN_SHOULD_SKIP_THIS // Doxygen misparses the below constructor, so dumb it down for it.
+  //! @brief Conversion from a \c basic_any_resource with the same set of properties but in a different order.
+  //! This constructor also handles conversion from \c async_any_resource to \c any_resource
+  //! @param __ref The other \c basic_any_resource.
```
Comment (bernhardmgruber) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.h:110]: Important: There is no parameter `__ref`, only `__other`.
```suggestion
  //! @param __other The other \c basic_any_resource.
```

File: cudax/include/cuda/experimental/__memory_resource/any_resource.h [hunk 2]
```diff
@@ -227,13 +244,33 @@ template <class _Ty, _CUDA_VMR::_AllocType _Alloc_type, class... _Properties>
...
-//! @brief Type erased wrapper around a `resource` that satisfies \tparam _Properties
-//! @tparam _Properties The properties that any resource wrapped within the `any_resource` needs to satisfy
+//! @rst
+//! .. _cudax-memory-resource-any-resource:
+//!
+//! Type erased wrapper around a `resource`
+//! ----------------------------------------
+//!
+//! ``any_resource`` wraps any given :ref:`resource <libcudacxx-extended-api-memory-resources-resource>` that
+//! satisfies the required properties. It owns the contained resource taking care of construction / destruction.
```
Comment (bernhardmgruber) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.h:254]:
```suggestion
//! satisfies the required properties. It owns the contained resource, taking care of construction / destruction.
```

File: cudax/include/cuda/experimental/__memory_resource/any_resource.h [hunk 3]
```diff
@@ -227,13 +244,33 @@ template <class _Ty, _CUDA_VMR::_AllocType _Alloc_type, class... _Properties>
...
-//! @brief Type erased wrapper around a `async_resource` that satisfies \tparam _Properties
-//! @tparam _Properties The properties that any async resource wrapped within the `async_any_resource` needs to satisfy
+//! @rst
+//! .. _cudax-memory-resource-async-any-resource:
+//!
+//! Type erased wrapper around an `async_resource`
+//! -----------------------------------------------
+//!
+//! ``async_any_resource`` wraps any given :ref:`async resource <libcudacxx-extended-api-memory-resources-resource>`
+//! that satisfies the required properties. It owns the contained resource taking care of construction / destruction.
```
Comment (bernhardmgruber) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.h:269]:
```suggestion
//! that satisfies the required properties. It owns the contained resource, taking care of construction / destruction.
```

**Review by bernhardmgruber — APPROVED** [review]

File: cudax/include/cuda/experimental/__memory_resource/any_resource.h [hunk 4]
```diff
@@ -53,6 +53,16 @@ namespace cuda::experimental::mr
...
 _LIBCUDACXX_INLINE_VAR constexpr bool __is_basic_any_resource = false;
 
+//! @rst
+//! .. _cudax-memory-resource-basic-any-resource:
+//!
+//! Base class for a type erased owning wrapper around a memory resource
+//! ---------------------------------------------------------------------
+//!
+//! ``basic_any_resource`` abstracts the differences between a resource and an async resource away, allowing efficient
+//! interoperability between the two.
```
Comment (ericniebler) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.h:63]: i'm not sure what this is referring to. `any_resource` and `any_async_resource` are different types with different interfaces. the differences haven  't been abstracted away.

File: cudax/include/cuda/experimental/__memory_resource/any_resource.h [hunk 5]
```diff
@@ -95,10 +105,12 @@ class basic_any_resource
     }
   }
 
-#ifndef DOXYGEN_SHOULD_SKIP_THIS // Doxygen misparses the below constructor, so dumb it down for it.
+  //! @brief Conversion from a \c basic_any_resource with the same set of properties but in a different order.
```
Comment (ericniebler) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.h:108]: i think we should require properties to have a UUID. that way we can have a canonical global ordering of properties. not for this PR obviously.

File: cudax/include/cuda/experimental/__memory_resource/any_resource.h [hunk 6]
```diff
@@ -95,10 +105,12 @@ class basic_any_resource
...
+  //! This constructor also handles conversion from \c async_any_resource to \c any_resource
+  //! @param __other The other \c basic_any_resource.
   _LIBCUDACXX_TEMPLATE(_CUDA_VMR::_AllocType _OtherAllocType, class... _OtherProperties)
   _LIBCUDACXX_REQUIRES(
-    _CUDA_VSTD::_IsNotSame<basic_any_resource, basic_any_resource<_OtherAllocType, _OtherProperties...>>::value
+    (_CUDA_VSTD::_IsNotSame<basic_any_resource, basic_any_resource<_OtherAllocType, _OtherProperties...>>::value)
```
Comment (ericniebler) [inline cudax/include/cuda/experimental/__memory_resource/any_resource.h:113]: ah, much better. thanks.

File: docs/repo.toml [hunk 7]
```diff
@@ -371,21 +371,19 @@ doxygen_predefined = [
...
   "_LIBCUDACXX_GLOBAL_CONSTANT=inline",
   "_LIBCUDACXX_INLINE_VAR=inline",
-  "_LIBCUDACXX_REQUIRES(...)=",
-  "_LIBCUDACXX_TEMPLATE(A)=template<A>",
-  "_LIBCUDACXX_TEMPLATE(A, B)=template<A, B>",
-  "_LIBCUDACXX_TEMPLATE(A, B, C)=template<A, B, C>",
-  "_LIBCUDACXX_TEMPLATE(A, B, C, D)=template<A, B, C, D>",
-  "_LIBCUDACXX_TEMPLATE(A, B, C, D, E)=template<A, B, C, D, E>",
+  "_LIBCUDACXX_REQUIRES(x)= ::cuda::std::__enable_if_t<x, int> = 0>",
+  "_LIBCUDACXX_TEMPLATE(x)=template<x, ",
```
Comment (ericniebler) [inline docs/repo.toml:386]: do we really want our docs to have the `enable_if` gunk?

**Review by ericniebler — COMMENTED** [review]
a few questions, otherwise looks good. thanks. i will be better about documenting my public additions.

**Review by ericniebler — APPROVED** [review]

File: docs/repo.toml (hunk 7 above)
Comment (miscco) [reply docs/repo.toml:386]: The issue is that doxygen does not support `requires` so we either have SFINAE or nothing
